src/lib.rs

```diff
@@ -383,6 +383,59 @@ impl<L, R> Either<L, R> {
     }
 }
 
+impl<T, L, R> Either<(T, L), (T, R)> {
+    /// Factor out a homogeneous type from an either of tuples.
+    ///
+    /// Here, the homogeneous type is the first element of the tuples.
+    ///
+    /// ```
+    /// use either::*;
+    /// let left: Either<_, (u32, String)> = Left((123, vec![0]));
+    /// assert_eq!(left.factor_first().0, 123);
+    ///
+    /// let right: Either<(u32, Vec<u8>), _> = Right((123, String::new()));
+    /// assert_eq!(right.factor_first().0, 123);
+    /// ```
+    pub fn factor_first(self) -> (T, Either<L, R>) {
+        self.either(|(t, l)| (t, Left(l)), |(t, r)| (t, Right(r)))
```

Comment:
I think these would be better as an explicit `match`.  It's slightly more verbose, but we only have to write it once, and it should be easier for the compiler to optimize.